File: 프로그래머스/2/138476. 귤 고르기/귤 고르기.java
import java.util.*;
import java.util.stream.Collectors;
class Solution {
    public int solution(int k, int[] tangerine) {
        int answer = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for(int t : tangerine){
            int cnt = map.getOrDefault(t, 0);
            map.put(t, cnt + 1);
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        for(int key : map.keySet()){
            pq.add(new int[]{key, map.get(key)});
        }
        k = tangerine.length - k;
        while(k > 0){
            int[] t = pq.poll();
//            System.out.println("t : " + t[0] + ", cnt : " + t[1]);
            k -= t[1];
            if(k < 0){
                pq.add(t);
            }
        }
        return pq.size();
    }
}

File: 프로그래머스/2/138476. 귤 고르기/귤 고르기.kt
import java.util.*
class Solution {
    fun solution(k: Int, tangerine: IntArray): Int {
        var answer: Int = 0
        var sum = 0
        // map으로 각 귤 크기 별 갯수를 센다
        // sort를 하는데 내림차순으로 해서 하나 씩 꺼내면서 값을 누적했을 때 k 만큼이 되면 그때까지의 종류를 카운팅한다.
        val countingMap = mutableMapOf<Int, Int>()
        for(t in tangerine) {
            countingMap[t] = countingMap.getOrDefault(t, 0) + 1
        }
        val entries = LinkedList(countingMap.entries)
        entries.sortBy { -it.value }
        for(e in entries){
            answer++
            sum += e.value
            if(sum >= k) {
                break
            }
        }
        return answer
    }
}